### ISSUE TITLE
Tracking issue for `associated_consts` feature

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#29646]
The feature is currently quite buggy, but this issue tracks its eventual stabilization.

- [ ] ~~Blocked on [`const_fn`](https://github.com/rust-lang/rust/issues/24111).~~
- [ ] ~~Type params in const expressions (#34344 / #39211)~~
- [x] Associated consts and object safety (#26847)
- [x] Do not always require `T: 'static` for `T::CONST`
- [x] Associated constants not fully checked between `trait` and `impl` (#41323)
- [x] Docs
  - https://github.com/rust-lang-nursery/reference/pull/75

### Associated consts in const expressions must be concrete

Associated consts can only be used in const contexts if their input types are _fully concrete_. That is, this works:

```rust
trait Foo {
     const X: usize;
}

impl Foo for () {
     const X = 0;
}

let y: [i32; <() as Foo>::X];
```

But this does not work:

```rust
fn foo<T: Foo>() {
    let y: [i32; <T as Foo>::X];
}
```

### Associated consts are never object safe

Associated consts make a trait non-object-safe. There is no way to bound the const `where Self: Sized`, nor to specify the const when creating the object.

```rust
trait Foo {
    const FOO: i32;
}

// Not allowed:
Box<Foo<FOO = 0>>

trait Bar {
    // Not allowed
    const BAR: i32 where Self: Sized;
}
```

### Associated consts referencing other associated consts have spurious destructor warnings

This works:

```rust
trait Foo {
    const FOO: Option<i32>;
}

trait Bar: Foo + 'static {
    const BAR: Option<i32> = <Self as Foo>::FOO;
}
```

But this does not, because we (unnecessarily) assume the const could be a variant that runs a destructor:

```rust
trait Foo {
    const FOO: Option<String>;
}

trait Bar: Foo + 'static {
    const BAR: Option<String> = <Self as Foo>::FOO;
}
```

**Comment by raindev** [comment]
I'm curious about the reasons behind this feature. Why aren't ordinary functions good enough for the same purpose? Given that Rust have default methods it shouldn't be much less ergonomic. And combination of monomorphisation and static lifetimes should converge any difference in terms of performance.

**Comment by Ketsuban** [comment]
One notable case is that of the `Zero` and `One` traits, which exist solely to confer their constant value to the types which implement them. With associated constants they would be redundant, as `Sum` and `Mul` would include the additive and multiplicative identities.

**Comment by raindev** [comment]
@Ketsuban how would it be different than including `zero` and `one` _as functions_ into `Sum` and `Mul` traits?

**Comment by raindev** [comment]
Having skimmed through #27739 it looks like there're opinions about the concepts of "zero" and "one" as functions as it may gives additional flexibility.

**Comment by mitaa** [comment]
@raindev
> Why aren't ordinary functions good enough for the same purpose?

https://github.com/rust-lang/rfcs/blob/master/text/0195-associated-items.md#user-content-expressiveness

**Comment by vberger** [comment]
The use of associated consts could make things like `bitflags!` more pretty in my opinion, by scoping the consts in the name of the bitflags struct, much like enum variants are scoped in the name of the enum.

**Comment by Roxxik** [comment]
would love to use associated consts for array sizes for example: http://is.gd/VSee6B

    struct Something {
        inner: [u8; Something::LENGTH]
    }

but this will leave me at E0250

**Comment by Robbepop** [comment]
> I'm curious about the reasons behind this feature. Why aren't ordinary functions good enough for the same purpose?

I am not sure about this but I think that associated constants could allow us compile time execution for things similar to non-type template parameters as in C++ which are quite useful. Using functions wouldn't work in the example below since Rust currently has no mechanism to detect if a function can be evaluated at compile-time (constexpr as in C++ or even templates).

**Example:**
https://gist.github.com/Robbepop/da73db721c22492616f6

If the example above really works (haven't tested it, yet) things like BitFlag<N> or the shown BigInt<N> could be implemented very efficiently - not using the heap at all!

**Comment by Boddlnagg** [comment]
Is there any ETA on the stabilization of this feature? It would be really useful for something which I'm currently experimenting with, namely the implementation of Windows COM interop: One could add a trait that exposes each interface's GUID and/or class ID as constants, then e.g. [this API](https://msdn.microsoft.com/en-us/library/windows/desktop/br224648%28v=vs.85%29.aspx) could be mapped to Rust in a way that takes the to-be-activated class as a type parameter that implements said trait.

Of course I can work on this using nightly, but it would be nice to have some information whether the feature is going to be stabilized in the near future!

**Comment by nrc** [comment]
Should we block on #34344? Are there other known bugs? Are there design questions to resolve here?

**Comment by letheed** [comment]
There was a question on the forum [here](https://users.rust-lang.org/t/why-does-this-code-need-a-static-bound-on-a-type-when-all-it-uses-is-const/7236). Looks like a bug to me.

**Comment by hauleth** [comment]
What is status or what would need to be done to stabilize this feature? It would be really helpful in Octavo as it would allow me to provide HMAC without any allocations.

**Comment by seanmonstar** [comment]
@hauleth take a look at the [associated-items issues](https://github.com/rust-lang/rust/labels/A-associated-items). That contains many of the bugs that exist with associated constants. I imagine many of them would need to be fixed before anyone could consider this feature "stable".

**Comment by brson** [comment]
How can we accelerate this? This is a major feature and it doesn't appear to be moving or have a champion.

This is [used by Rocket](https://github.com/SergioBenitez/Rocket/issues/19#issuecomment-269052006).

**Comment by brson** [comment]
Just nominating to put it in front of eyeballs again.

**Comment by steveklabnik** [comment]
My understanding is that this is waiting on a re-do of const evaluation in the compiler, since it's so buggy. @eddyb ?

**Comment by eddyb** [comment]
"Buggy" is not the best way to put it IMO, "barely does anything at all" is far closer.
@brson You can consider @solson, @oli-obk (both miri) and me (rustc) as unofficial champions (for the past several months) for const-eval-related areas.

**Comment by brson** [comment]
Noted in the op that this is blocked on const_fn.

**Comment by brson** [comment]
Should block on this bug https://github.com/rust-lang/rust/issues/26847

**Comment by JoeyAcc** [comment]
Out of curiosity, would something like the snippet below work in a recent Rust version? I'm thinking either the most recent stable, or straight up nightly.

````
pub struct Foo(/* elided */);

impl Foo {
    pub const MY_CONST: &'static str = "hurr durr I'm a sheep";
    // Other defs elided
}
````

**Comment by hauleth** [comment]
@JoeyAcc if you need to use `feature` then 100% it will not work on anything other than nightly.

**Comment by JoeyAcc** [comment]
Yes, I understand that. It wasn't quite what I intended to ask.
Rather, I wanted to know if the code above even works at all, since there are apparently still issues with the feature *somewhere* (and I have no clue where exactly the breakage is currently).
It's more about the line (or some rough approximation of it) than which Rust version I should use.
Sorry about the confusion.

**Comment by sagebind** [comment]
I'm able to get away with using this instead of `const fn` for certain types of "constructors" in order to yield a better API. For example:

```
pub struct SomeType<T> {
    // Some fields that are intentionally private...
}

impl<T> SomeType<T> {
    pub const DEFAULT: Self = Self {
        // Initialize fields...
    };
}

// Now we can create new `SomeType<T>` values generic over T without a function call...
let value: SomeType<i32> = SomeType::DEFAULT;
```

**Comment by withoutboats** [comment]
Can anyone give an update on associated consts? @eddyb ?

The tracking issue here lists two items: const fn & docs. Setting aside docs, I don't know why const fn is a necessary precursor to associated consts. It seems to me that associated consts could instead be the beachhead which const fn will someday make more useful.

I also suspect the issue is incomplete; in the past there were many bugs with this feature. While when I've used it more recently its seemed better, but I see some open bugs linked from here.

Personally, I'd like to have this on stable not because I *necessarily* care about compile time computation, but also because I have a lot of these `fn whatever() -> &'static str;` methods that feel more correct & ergonomic as `const WHATEVER: &'static str;`.

I don't think there are any design issues with this feature left, and I think we are certain that we want it in the language. I wonder how close it is to a point where we could stabilize *something*.

**Comment by eddyb** [comment]
@withoutboats I think that was because I told @brson the two features are related.
However, after #40008, everything except type parameters in array lengths should work.
We could even stsbilize the feature as it is right now, and your usecase would be covered.

**Comment by withoutboats** [comment]
@eddyb That's great to hear. I triaged the open issues linked from this thread & I agree that we could stabilize this feature in the short term. Maybe someone else has reason not to, but there's an easy way to hear from them...

@rfcbot fcp merge

@rust-lang/lang I propose we stabilize associated consts. :tada:

On their own they're not that *useful* (because they're primarily *useful* as the input to constexprs, of which array lengths are the only stable example), but they have a more "para-semantic" use as a more precise way to express oneself.

That is, today I might write:

```rust
trait Whatever {
    fn name() -> &'static str;
}
```

But what I really want to write is:

```rust
trait Whatever {
    const NAME: &'static str;
}
```

This is both more ergonomic & it more clearly expresses *user intent* than the associated function form.

There are currently two outstanding shortcomings tracked:

* __Type params in const expressions:__ ( #34344 / #39211 ) A const expression (such as an array length) cannot today depend on an abstract type parameter. That is `for<T: Trait> [_; T::CONST]` is not usable; it returns a clean error. However, `[_; i32::CONST]` _does_ work. In my stabilization need not block this. **Proposal: No work needed.**

* __Associated consts and object safety:__ ( #26847) Today, associated consts are considered object safe, but if you attempt to access the associated const of a trait object type (as in `<Trait as Trait>::CONST`) you get an ICE. I think the most conservative thing to do is to make associated consts (like associated functions) non-object safe. **Proposal: make associated consts not object safe; block stabilization on this.**

**Comment by SimonSapin** [comment]
> A const expression (such as an array length) cannot today depend on an abstract type parameter. That is `for<T: Trait> [_; T::CONST]` is not usable; it returns a clean error.

Can the language be extended to support this in the future? Is it a matter of compiler implementation work?

**Comment by eddyb** [comment]
@SimonSapin It's more or less the same as getting const generics to work, although a bit easier.
I don't think that limitation is specified anywhere, it's more of an implementation detail that we hope to fix.

**Comment by withoutboats** [comment]
Does `[_; T::CONST]` raise the sort of unification questions etc that the const params RFC does? If so, if its implemented before thats resolved it should be feature gated.

**Comment by seanmonstar** [comment]
Regarding object safety, would there be an implicit 'where Self: Sized' constraints on constants (since we can't write them ourselves)? Or would it just be illegal to box a trait object that has constants? 

I'd prefer the former, and have a specific instance that has been waiting on this: `Header` in hyper.

**Comment by eddyb** [comment]
@withoutboats The conservative approach (treating it the same way we would a `T::Assoc` type, i.e. a projection) is backwards-compatible, and everything else after that is "peeking through" projections.

**Comment by withoutboats** [comment]
@eddyb I don't understand what you mean. What does "treating it the same way we would a [projection]" look like to users? Its backwards compatible with what?

**Comment by eddyb** [comment]
@withoutboats `T::Assoc` does not determine `T`, so `T` cannot appear *solely* in projections in an `impl` header, i.e. `impl<T: Trait> Foo for T::Assoc` gives an error that `T` is unconstrained.
Also, `T::Assoc` is considered to overlap with *any other type* (as it could be that type), for coherence.

`[U; T::CONST]` should have both of those behaviors, at least initially, and that is backwards-compatible with *any other scheme* (at least those that I know of/can think of right now), as it's the most conservative.

**Comment by withoutboats** [comment]
@eddyb Oh I see, you're responding to the unification question not the object safety question. Sounds good. :-)

**Comment by withoutboats** [comment]
@seanmonstar The proposal I made is for associated consts to simply not be object safe. Your example is a good one; I wouldn't want to implicitly add `where Self: Sized` to these items, but I think we should add the ability to attach where clauses to them to solve this problem, just like functions have. I don't know that we should block on that though.

**Comment by nrc** [comment]
How long have assoc consts been implemented? I was not aware they were ready for testing (only that they were sort of done/wip). I think we should perhaps advertise that they are usable and then wait a few weeks before stabilising. It seems a bit rushed to stabilise a big feature like this that is not getting much use.

**Comment by withoutboats** [comment]
I've occassionally turned this feature on when I wanted it to express intent. I've always ended up turning it off because it wasn't worth blocking stable on.

My anecdotal experience is that in 2015, when I tried these were just totally unusable (like I think even in runtime expressions you couldn't do `T::CONST`). When I tried again in 2016, I found that they work as you'd expect in a runtime context. The only thing that's been implemented recently is that concrete associated consts (`i32::CONST`) can now be used in constexpr context.

So yea for about a year now my experience with this feature has been that it works, but its not been worth moving off stable for.

Rocket apparently uses this feature, maybe someone involved in that project can give feedback on whether there have been any issues.

**Comment by gnzlbg** [comment]
Things like these: https://play.rust-lang.org/?gist=052e577b29c316cbb06a9584f6006404&version=nightly&backtrace=0

still do not work, is that normal? (I know they can be worked around, but the error message could at least use some love).

**Comment by withoutboats** [comment]
Yea that's the same as `[_; T::CONST]`, its a const based on a non-concrete type.

**Comment by crumblingstatue** [comment]
[bitflags](https://github.com/rust-lang-nursery/bitflags) could benefit from this feature. See:
- https://github.com/rust-lang-nursery/bitflags/issues/20
- https://github.com/rust-lang-nursery/bitflags/issues/21

It would improve discoverability if for `fn foo(flags: Bar)` one could pass `foo(Bar::BAZ)` instead of `foo(BAZ)`.

**Comment by burdges** [comment]
I donno if [this](https://is.gd/gjDbpN) comes under https://github.com/rust-lang/rust/issues/34344 but it throws _error[E0311]: the parameter type `P` may not live long enough_.

If we want to copy a `const` between trait `impl`s, then we need a `'static` bound on either `P` or `Self`, which seems overly restrictive. 
```
pub trait Params: Sized {
    /// Unique version identifier for the protocol
    const PROTOCOL_NAME: u64;
}

pub struct ParamsEtc<P: Params>(P);
impl<P> Params for ParamsEtc<P> where P: Params {
    const PROTOCOL_NAME: u64
     = P::PROTOCOL_NAME;
}
```

**Comment by eddyb** [comment]
@burdges Yeah that's sort of a bug. It's a limitation in the current regionck. cc @nikomatsakis

**Comment by eddyb** [comment]
@aturon @withoutboats I've updated the issue description based on comments here (and to add #41323).

**Comment by withoutboats** [comment]
@aturon @nikomatsakis @pnkfelix Are there any concerns about stabilizing this feature?

**Comment by nikomatsakis** [comment]
@withoutboats I do have concerns, but probably they don't have to block stabilization. Primarily, I am concerned that the existing impl is "ad hoc" in various ways and that we may discover surprising things that it allows or disallows. I'm not sure how best to allay those concerns (and they apply to a lot of things that are stable anyway).

**Comment by withoutboats** [comment]
> Primarily, I am concerned that the existing impl is "ad hoc" in various ways and that we may discover surprising things that it allows or disallows.

Yes, and I've run into these since turning it on, though they've all been things where the current implementation is too conservative.

For example, this works, but requires a `'static` bound that's definitely unnecessary:

```rust
trait Foo {
    const FOO: u32;
}

trait Bar {
    const BAR: u32;
}

impl<T: Foo + 'static> Bar for T {
    const BAR: u32 = T::FOO;
}
```

This doesn't work, even though it should:

```rust
// This trait works:
trait Foo {
    const FOO: Option<String>;
}

trait Bar {
    const BAR: Option<String>;
}

impl<T: Foo + 'static> Bar for T {
    const BAR: Option<String> = T::FOO;
}
```

However, where this does work its strictly more expressive than you might expect from the errors you get for `[T; T::LEN]`, even if its strictly less expressive than it ought to be. It seems jagged and surprising but ultimately implementation bugs & I think that it doesn't have to block stabilization but YMMV.

**Comment by aturon** [comment]
@withoutboats I'm fine signing off on the top-level design, but like @nikomatsakis feel uncertain about the state of the implementation. I'm not really sure how to gain confidence about it, though. In any case, I won't block stabilization.

**Comment by withoutboats** [comment]
There's an abstract question here: how can implementation issues block stabilization? (several other features we're pretty comfortable with the design of have similar issues - impl trait and specialization come to mind).

Here are the kinds of implementation issues that can crop up:

* Is the implementation *unsound*? Does it allow unsafety in safe code?
* Is the implementation *overly liberal*? Does it allow safe things we don't want it to allow?
* Is the implementation *overly conservative*? Does it disallow things we think should be allowed?

These seem like progressively less serious as you go down the list. Another axis each wrongness in the implementation needs to be considered on is its egregiousness.

For example, I think we should block stabilizing this the fact that consts are considered object safe today. This is both too liberal and very egregious.

But that we don't accept some uses of associated consts in other associated consts is both too conservative (not as big an issue) and not particularly egregious, and I'm comfortable not blocking on that.

**Comment by aturon** [comment]
@withoutboats I think what this rubric comes down to is: how likely is it that we will need/want to make breaking changes to the implementation in the future?

I'm not at all expert in our const system, so I'm basically happy to trust that those that are have sussed out the points of overly-liberal or overly-conservative behavior.

**Comment by arielb1** [comment]
@withoutboats 

I don't want to expose the current associated const system to user code because it *will* break horribly (well, slightly less horribly than a few months ago, but still). We are going to integrate MIRI with rustc soon-ish, and that should bring things to a saner state.

**Comment by withoutboats** [comment]
> I don't want to expose the current associated const system to user code because it will break horribly

Can you be more expansive?

**Comment by nikomatsakis** [comment]
After some discussion, I feel reasonably good about putting this into FCP for stabilization, but I would feel better if we could assemble:

- A list of known shortcomings, linked to from the top of this bug
- Clear links to the tests (ideally, centralized or tagged in some way) showing all the things that work as expected (or don't, as the case may be)

@withoutboats has promised to do some of this. For the record, I think I'd like this all the time as we stabilize major features (or maybe even minor ones).

**Comment by eddyb** [comment]
Just so I don't forget, this doesn't work right now ([playpen](https://play.rust-lang.org/?gist=5a332ddf940e55c4f8a3f9be157857c2&version=nightly&backtrace=1)):
```rust
trait Foo {
    const X: i32 = 0;
}

trait Bar: Foo {
    // error: cannot use an outer type parameter in this context
    const Y: i32 = Self::X;
}
```
IIRC on the branch I started working towards `[T; T::N]` fixes this, I'll extract that into a PR.

**Comment by JoeyAcc** [comment]
It wasn't entirely obvious to me what would happen, so I also just tried using `Foo` rather than `Self`, like so:

```` rust
trait Foo {
    const X: i32 = 0;
}

trait Bar: Foo {
    // error: cannot use an outer type parameter in this context
    const Y: i32 = Foo::X;
}
````

The result in the playpen is the same as if `Self` had been used.

**Comment by eddyb** [comment]
@JoeyAcc I don't see how that's possible. I get:
```rust
error[E0283]: type annotations required: cannot resolve `_: Foo`
  --> <anon>:18:20
   |
18 |     const Y: i32 = Foo::X;
   |                    ^^^^^^
   |
   = note: required by `Foo::X`
```

**Comment by JoeyAcc** [comment]
Ah yes I see. It fails with a completely different error, and that fact slipped by me.
Oops, carry on :)

**Comment by withoutboats** [comment]
@nikomatsakis I've updated the first post to enumerate the shortcomings of this feature that I know about. I also checked the `run-pass` tests and they seemed comprehensive to me.

**Comment by eddyb** [comment]
@withoutboats Isn't "Types referenced in associated consts must be 'static" solved already?

**Comment by withoutboats** [comment]
Yes! I didn't see you'd fixed that, I thought you just fixed the using `Self` in default associated consts.

There's a run-pass test that needs to be updated then, I'll PR..

**Comment by brson** [comment]
Will this make it in for 1.19?

**Comment by withoutboats** [comment]
I was hoping so but we just discovered a soundness bug at #41323, not sure if that will be fixed in time

**Comment by seanmonstar** [comment]
The soundness bug #41323 has been closed. Anything else preventing stabilizing?

**Comment by withoutboats** [comment]
Still needs docs. Also the cut off is today. I think we're looking at 1.20 now but maybe we could squeeze it through? Personally unclear about the policy on this.

**Comment by steveklabnik** [comment]
https://doc.rust-lang.org/nightly/unstable-book/language-features/associated-consts.html has the basics, would need to be adapted for the reference/book.

**Comment by brson** [comment]
Looks like it'll be 1.20.

**Comment by seanmonstar** [comment]
If I wanted to get this moving to stable now, are those docs sufficient? Would I just send a PR removing all the feature gate stuff?

**Comment by steveklabnik** [comment]
> If I wanted to get this moving to stable now, are those docs sufficient?

Sending in a PR for them would be, yeah. We might iterate a bit in the PR, but the docs requirement is more of a "a PR must be sent in for docs" than "the docs must be merged"

**Comment by shepmaster** [comment]
The original comment has #34344 and #39211 listed as a blocker and both of those are open, although with comments along the lines of "this probably works". I'd be nice to have those resolved concretely one way or the other.

**Comment by eddyb** [comment]
@shepmaster We decided those wouldn't block stabilization of associated consts.

**Comment by shepmaster** [comment]
> We decided those wouldn't block stabilization of associated consts.

@eddyb cool. I've edited the comment to strike those out like the no-longer-blocking `const fn`.

**Comment by nikomatsakis** [comment]
So we have current stabilized some part of the existing feature in https://github.com/rust-lang/rust/pull/42809 (and there are docs). I think it'd be really great if we updated the description with a more comprehensive write-up of what *does work* and what *does not* -- though the existing listing of shortcomings seems like a good start.

**Comment by mikeyhew** [comment]
Does the stabilization of this feature mean that generic fixed-sized arrays will be supported soon?

**Comment by SimonSapin** [comment]
@mikeyhew The original message cites `let y: [i32; <T as Foo>::X];` explicitly an something that does **not** work. And even if it did, that still doesn’t give integer generics.

**Comment by withoutboats** [comment]
@mikeyhew You can follow the progress on that feature on this rfc: rust-lang/rfcs#2000

**Comment by withoutboats** [comment]
@nikomatsakis I think the list of shortcomings is an accurate summary of what doesn't work. What does work is "everything else," isn't it? What is missing from the tracking issue that isn't covered in the new docs for the feature?

**Comment by eddyb** [comment]
@mikeyhew That is a different feature, although it does depend on lifting the limitations of array lengths not being able to depend on generic parameters, that were punted on for the stabilization of associated consts (see earlier discussion).

**Comment by Object905** [comment]
Today found that lifetime in associated consts doesn't default to `'static` as according to [stabilized in 1.17.0](https://github.com/rust-lang/rust/pull/39265) [rfc](https://github.com/rust-lang/rfcs/blob/master/text/1623-static.md). So: 
`trait Foo { const FOO: &str = "dsad"; }` gives `expected lifetime parameter` error, while `trait Foo { const FOO: &'static str = "dsad"; }` is valid. Same goes for struct, enum and union impl blocks.

**Comment by frewsxcv** [comment]
@Object905 Relevant discussion in https://github.com/rust-lang/rust/issues/38831

**Comment by steveklabnik** [comment]
Should this be closed? All boxes are checked off, and https://github.com/rust-lang/rust/pull/42809 was merged

**Comment by shepmaster** [comment]
And it's available in beta AFAICT

**Comment by steveklabnik** [comment]
Yes, it will be in 1.20.

**Comment by nwin** [comment]
Maybe I’m not seeing the point but as long as “Associated consts in const expressions must be concrete” is still a restriction I do not see any use of associated constants in traits. 

Yet you can add them such that all of the error messaged presented in #32344 are still very there and very misleading.

Shouldn’t one simply forbid associated constants in traits unless that restriction is lifted? I dont’t think that it was a good idea merge #42809 (i.e. remove the feature gate) with a feature in such an incomplete state.

**Comment by eddyb** [comment]
@nwin It's not "in const expressions" - but rather "in types", i.e. array types.

**Comment by frehberg** [comment]
Is there any plan to introduce LEN (or similar attribute) as associated const for the built-in rust-primitives 'arrray' or 'slice'? , for example 

const N : usize = b"my_byte_string"::LEN;

or 

is there any plan to declare std::intrinsics::size_of_val(v)  as const-fn(?) enabling somehing like:
#![feature(core_intrinsics)]
const N: usize = unsafe { std::intrinsics::size_of_val(b"my_byte_string") };

**Comment by eddyb** [comment]
@frehberg When const generics are implemented you will be able to do that yourself, for example with an `impl<T, const N: usize> MyLenTrait for [T; N]`

**Comment by robinst** [comment]
Anyone landing here from googling "associated const", this was released in Rust 1.20, see the announcement: https://blog.rust-lang.org/2017/08/31/Rust-1.20.html

**Comment by robinst** [comment]
Also, if anyone is looking for the documentation of this, it's here: https://doc.rust-lang.org/stable/reference/items.html#associated-constants

(It was surprisingly hard to find.)

**Comment by SergioBenitez** [comment]
Shouldn't equality bounds be possible with associated consts? The following does not compile today:

```rust
trait Foo {
    const BAR: usize;
}

// error: expected type, found `10`
fn f<T: Foo<BAR = 10>>() { }
```

The original [associated items RFC](https://github.com/rust-lang/rfcs/blob/master/text/0195-associated-items.md#constraining-associated-types) makes it clear that this should be possible. I'm wondering: was this simply forgotten about? Or perhaps there are/were complications with implementing this?

**Comment by SergioBenitez** [comment]
Furthermore, using an associated const as an array size does not appear to work:

```rust
pub trait Can {
    const SIZE: usize;
}

fn f<T: Can>(t: T) {
    // error[E0277]: the trait bound `T: Can` is not satisfied
    let x = [0u8; <T as Can>::SIZE];
}
```

There error message is clearly wrong, so this is either a bug in `rustc` or unimplemented functionality resulting in a bogus error message.

**Comment by shepmaster** [comment]
> using an associated const as an array size

Probably related to #42863, #44168, #44774 ?

**Comment by gnzlbg** [comment]
> Furthermore, using an associated const as an array size does not appear to work:

This usage of associated const was not stabilized, as documented in this issue (it doesn't work on nightly either, and probably won't work until `const`-generics implementation improves).

**Comment by gnzlbg** [comment]
I don't know if this is the same issue that @eddyb mentioned here: https://github.com/rust-lang/rust/issues/29646#issuecomment-300929548

```rust
trait Foo { const A: usize; }
fn foo<F: Foo>(f: F) {
    const B: usize = F::A; // ERROR
}
```

**Comment by eddyb** [comment]
@gnzlbg Those are two items nested in eachother, we don't allow "inheriting" type parameters in such a situation and it'd be a significant language change to start doing that.
IOW, what you wrote is more or less equivalent to:
```rust
const __HIDDEN_B: usize = F::A;
fn foo<F: Foo>(f: F) {
    use self::__HIDDEN_B as B;
}
```
(and this is not in any way, shape, or form, specific to `const` or associated `const`)

**Comment by zpgaal** [comment]
Having const fn is this feature still required ? Of course it'd require to have const fn in traits as well.
(For now no array size can be given using either associated const or const fn)

**Comment by steveklabnik** [comment]
@zpgaal this feature is implemented and stable, so regardless of it being "required", it has to stay!

**Comment by vks** [comment]
It seems like rustdoc does not generate documentation for associated constants in structs. Is this a bug?

**Comment by SimonSapin** [comment]
@vks I cannot reproduce this. `rustdoc +stable a.rs` with the code below does document the constant.

```rust
pub struct Foo;

impl Foo {
    pub const BAR: u32 = 7;
}
```

**Comment by gnzlbg** [comment]
@SimonSapin I think the problem is: 

```rust
pub struct Foo;

impl Foo {
    /// This doc comment won't be shown anywhere.
    pub const BAR: u32 = 7;
}
```

**Comment by SimonSapin** [comment]
Oh I see. Please file a bug separately?

**Comment by vks** [comment]
> Oh I see. Please file a bug separately?

See #52260.
